### PR TITLE
Add loadbalancer support

### PR DESCRIPTION
**Pull request by mrjana**
This PR adds support for loadbalancing across a group of endpoints that
share the same service configuration as passed in by
`OptionService`. The loadbalancer is implemented using ipvs with just
round robin scheduling supported for now.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

### CODE REVIEW

File: endpoint.go [hunk 1]
```diff
@@ -93,6 +94,7 @@ func (ep *endpoint) MarshalJSON() ([]byte, error) {
 	epMap["myAliases"] = ep.myAliases
 	epMap["svcName"] = ep.svcName
 	epMap["svcID"] = ep.svcID
+	epMap["virtualIP"] = ep.virtualIP.String()
```
Comment (aboch) [inline endpoint.go]: Shouldn't we add it to the map only if present, if not nil ?
IP.String() may handle a nil IP, but I am not sure how the unmarshal is going to behave.

File: service.go [hunk 2]
```diff
@@ -69,25 +151,255 @@ func (c *controller) rmServiceBinding(name, sid, nid, eid string, ip net.IP) err
...
+	}
+}
+
+// Add loadbalancer backend to all sandboxes which has a connection to
+// this network. If needed add the service as well, as specified by
+// the addService bool.
+func (n *network) addLBBackend(ip, vip net.IP, fwMark uint32, addService bool) {
+	n.WalkEndpoints(func(e Endpoint) bool {
+		ep := e.(*endpoint)
+		sb, ok := ep.getSandbox()
```
Comment (aboch) [inline service.go]: Minor: 
Given the walk function will never interrupt the walk, we could rewrite it as follow, it also seems more in line with the goal

```
func(e Endpoint) bool {
    ep := e.(*endpoint)
    if sb, ok := ep.getSandbox(); ok {
        sb.addLBBackend(ip, vip, fwMark, addService)
    }
    return false
}
```
Same in `rmLBBackend()`

File: service.go [hunk 3]
```diff
@@ -69,25 +151,255 @@ func (c *controller) rmServiceBinding(name, sid, nid, eid string, ip net.IP) err
...
+
+	vip := os.Args[2]
+	fwMark, err := strconv.ParseUint(os.Args[3], 10, 32)
+	if err != nil {
+		logrus.Errorf("bad fwmark value(%s) passed: %v", os.Args[3], err)
+		os.Exit(2)
+	}
+	addDelOpt := os.Args[4]
+
+	f, err := os.OpenFile(os.Args[1], os.O_RDONLY, 0)
```
Comment (aboch) [inline service.go]: Any reason for not using `netns.GetFromPath()` instead. I thought it would be equivalent.

File: agent.go [hunk 4]
```diff
@@ -167,14 +167,17 @@ func (ep *endpoint) addToCluster() error {
...
+			if err := c.addServiceBinding(ep.svcName, ep.svcID, n.ID(), ep.ID(), ep.virtualIP, ep.Iface().Address().IP); err != nil {
+				return err
+			}
 		}
 
 		buf, err := proto.Marshal(&EndpointRecord{
 			Name:        ep.Name(),
 			ServiceName: ep.svcName,
 			ServiceID:   ep.svcID,
+			VirtualIP:   ep.virtualIP.String(),
```
Comment (aboch) [inline agent.go]: Only if `ep.VirtualIP` can be nil, would the unmarshal properly reconstruct a nil pointer from a "\<nil>" string which we are passing here ?

File: endpoint.go (hunk 1 above)
Comment (mrjana) [inline endpoint.go]: `net.ParseIP` in unmarshal will return `nil` if the address sting is invalid.

File: service.go (hunk 2 above)
Comment (mrjana) [inline service.go]: Good suggestion. Will change.

File: service.go (hunk 3 above)
Comment (mrjana) [inline service.go]: Yeah forgot about that. Will change to use it.

**Comment by aboch** [comment]
LGTM

**Comment by mavenugo** [comment]
LGTM

**Comment by chenchun** [comment]
Gelivable!

**Pull request merged**